File: backend/analytics/core_signals.py
import pandas as pd
from .indicators import calculate_rsi, calculate_macd, calculate_ema

def generate_rsi_macd_signal(ohlcv_data: pd.DataFrame) -> dict:
    """Generate core RSI+MACD signal (40% weight in final algorithm)"""
    if len(ohlcv_data) < 50:
        return {
            'action': 'HOLD',
            'confidence': 0.3,
            'score': 0.5,
            'rsi_value': 50,
            'macd_histogram': 0,
            'strength': 0,
            'details': {}
        }
    
    prices = ohlcv_data['close']
    
    rsi = calculate_rsi(prices)
    macd_data = calculate_macd(prices)
    
    current_rsi = rsi.iloc[-1]
    current_macd_hist = macd_data['histogram'].iloc[-1]
    current_macd_line = macd_data['macd_line'].iloc[-1]
    current_signal_line = macd_data['signal_line'].iloc[-1]
    
    # Core trading logic (40% weight in final algorithm)
    action = 'HOLD'
    confidence = 0.3
    strength = 0
    
    if current_rsi < 30 and current_macd_hist > 0:
        action = 'BUY'
        confidence = 0.8
        strength = abs(current_macd_hist)
    elif current_rsi > 70 and current_macd_hist < 0:
        action = 'SELL'
        confidence = 0.8
        strength = abs(current_macd_hist)
    elif current_rsi < 40 and current_macd_hist > 0:
        action = 'BUY'
        confidence = 0.6
        strength = abs(current_macd_hist) * 0.7
    elif current_rsi > 60 and current_macd_hist < 0:
        action = 'SELL'
        confidence = 0.6
        strength = abs(current_macd_hist) * 0.7
    
    # Calculate normalized score (0-1)
    rsi_score = confidence if action in ['BUY', 'SELL'] else 0.5
    
    return {
        'action': action,
        'confidence': confidence,
        'score': rsi_score,
        'rsi_value': current_rsi,
        'macd_histogram': current_macd_hist,
        'strength': strength,
        'details': {
            'rsi': current_rsi,
            'macd_line': current_macd_line,
            'signal_line': current_signal_line,
            'histogram': current_macd_hist
        }
    }

def calculate_trend_strength(ohlcv_data: pd.DataFrame) -> float:
    """Calculate trend strength using EMAs"""
    if len(ohlcv_data) < 50:
        return 0.5
    
    prices = ohlcv_data['close']
    ema_20 = calculate_ema(prices, 20)
    ema_50 = calculate_ema(prices, 50)
    
    current_price = prices.iloc[-1]
    current_ema_20 = ema_20.iloc[-1]
    current_ema_50 = ema_50.iloc[-1]
    
    if current_price > current_ema_20 > current_ema_50:
        return 0.8  # Strong uptrend
    elif current_price < current_ema_20 < current_ema_50:
        return 0.8  # Strong downtrend
    else:
        return 0.4  # Sideways/weak trend

File: src/analytics/coreSignals.ts
import { technicalIndicators } from './indicators';
import { OHLCVData } from '../types';

export class CoreSignalAnalyzer {
  
  generateRSIMACDSignal(ohlcvData: OHLCVData[]): any {
    if (ohlcvData.length < 50) {
      return {
        action: 'HOLD',
        confidence: 0.3,
        score: 0.5,
        rsi_value: 50,
        macd_histogram: 0,
        strength: 0
      };
    }

    const prices = ohlcvData.map(d => d.close);
    const rsi = technicalIndicators.calculateRSI(prices);
    const macd = technicalIndicators.calculateMACD(prices);
    
    const currentRSI = rsi[rsi.length - 1];
    const currentMACDHist = macd.histogram[macd.histogram.length - 1];
    
    let action = 'HOLD';
    let confidence = 0.3;
    let strength = 0;
    
    // Core trading logic (40% weight in final algorithm)
    if (currentRSI < 30 && currentMACDHist > 0) {
      action = 'BUY';
      confidence = 0.8;
      strength = Math.abs(currentMACDHist);
    } else if (currentRSI > 70 && currentMACDHist < 0) {
      action = 'SELL';
      confidence = 0.8;
      strength = Math.abs(currentMACDHist);
    } else if (currentRSI < 40 && currentMACDHist > 0) {
      action = 'BUY';
      confidence = 0.6;
      strength = Math.abs(currentMACDHist) * 0.7;
    } else if (currentRSI > 60 && currentMACDHist < 0) {
      action = 'SELL';
      confidence = 0.6;
      strength = Math.abs(currentMACDHist) * 0.7;
    }
    
    // Calculate normalized score (0-1)
    const rsiScore = action === 'BUY' || action === 'SELL' ? confidence : 0.5;
    
    return {
      action,
      confidence,
      score: rsiScore,
      rsi_value: currentRSI,
      macd_histogram: currentMACDHist,
      strength,
      details: {
        rsi: currentRSI,
        macd_line: macd.macd_line[macd.macd_line.length - 1],
        signal_line: macd.signal_line[macd.signal_line.length - 1],
        histogram: currentMACDHist
      }
    };
  }

  calculateTrendStrength(ohlcvData: OHLCVData[]): number {
    if (ohlcvData.length < 50) return 0.5;
    
    const prices = ohlcvData.map(d => d.close);
    const ema20 = technicalIndicators.calculateEMA(prices, 20);
    const ema50 = technicalIndicators.calculateEMA(prices, 50);
    
    const currentPrice = prices[prices.length - 1];
    const currentEMA20 = ema20[ema20.length - 1];
    const currentEMA50 = ema50[ema50.length - 1];
    
    if (currentPrice > currentEMA20 && currentEMA20 > currentEMA50) {
      return 0.8; // Strong uptrend
    } else if (currentPrice < currentEMA20 && currentEMA20 < currentEMA50) {
      return 0.8; // Strong downtrend
    } else {
      return 0.4; // Sideways/weak trend
    }
  }
}

export const coreSignalAnalyzer = new CoreSignalAnalyzer();